File: src/main/python/org/o3project/odenos/remoteobject/remote_object.py
# -*- coding:utf-8 -*-

# Copyright 2015 NEC Corporation.                                          #
#                                                                          #
# Licensed under the Apache License, Version 2.0 (the "License");          #
# you may not use this file except in compliance with the License.         #
# You may obtain a copy of the License at                                  #
#                                                                          #
#   http://www.apache.org/licenses/LICENSE-2.0                             #
#                                                                          #
# Unless required by applicable law or agreed to in writing, software      #
# distributed under the License is distributed on an "AS IS" BASIS,        #
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. #
# See the License for the specific language governing permissions and      #
# limitations under the License.                                           #


import logging
import traceback
import copy

from org.o3project.odenos.remoteobject.object_property import ObjectProperty
from org.o3project.odenos.remoteobject.message.request import Request
from org.o3project.odenos.remoteobject.message.response import Response
from org.o3project.odenos.remoteobject.message.event import Event
from org.o3project.odenos.remoteobject.event.object_setting_changed import (
    ObjectSettingChanged
)
from org.o3project.odenos.remoteobject.manager.event_subscription import EventSubscription
from org.o3project.odenos.core.util.request_parser import RequestParser


class RemoteObject(object):
    DESCRIPTION = ""
    CONNECTION_TYPES = ""

    def __init__(self, arg1, arg2, arg3=None):
        if arg3 is None:
            object_id = arg1
            dispatcher = arg2
        else:
            object_id = arg1
            baseUrl = arg2
            dispatcher = arg3
        self._object_property = ObjectProperty(self.__class__.__name__,
                                               object_id)
        self._object_property.set_state(ObjectProperty.State.INITIALIZING)
        self.__set_description()
        self.__set_connection_types()
        self._object_settings = {}

        self.dispatcher = dispatcher
        if self.dispatcher is None:
            return
        self.dispatcher.add_local_object(self)
        self._event_subscription = EventSubscription(object_id)
        self.__parser = RequestParser()
        self.__add_rules()

    def on_initialize(self, obj_prop):
        return True

    def on_finalize(self):
        logging.debug("on_finalize ObjectID:" + self.object_id)
        self.dispatcher.remove_local_object(self)

    @property
    def object_id(self):
        return self._object_property.object_id

    @property
    def object_property(self):
        return self._object_property

    @property
    def object_settings(self):
        return self._object_settings

    @property
    def state(self):
        return self._object_property.state

    def set_state(self, state):
        prop = self.object_property
        prop.set_state(state)
        return self._do_put_property(prop.packed_object())

    def __set_description(self):
        self._object_property.set_property(ObjectProperty.DESCRIPTION,
                                           self.DESCRIPTION)

    def __set_connection_types(self):
        self._object_property.set_property(ObjectProperty.CONNECTION_TYPES,
                                           self.CONNECTION_TYPES)

    def _request_sync(self, object_id, method, path, body=None):
        return self.dispatcher.request_sync(Request(object_id,
                                                    method,
                                                    path,
                                                    body))

    def _request(self, object_id, method, path, body=None):
        resp = Response(Response.StatusCode.INTERNAL_SERVER_ERROR, None)
        try:
            resp = self._request_sync(object_id, method, path, body)
        except:
            logging.error("Exception: Request to " + object_id
                          + " Method:" + method
                          + " Path:" + path)
            logging.error(traceback.format_exc())

        return resp

    def _publish_event_async(self, event_type, body):
        return self.dispatcher.publish_event_async(Event(self.object_id,
                                                         event_type,
                                                         body))

    def _apply_event_subscription(self):
        return self.dispatcher.subscribe_event(self._event_subscription)

    def __add_rules(self):
        rules = []
        rules.append({RequestParser.PATTERN: r"^property/?$",
                      RequestParser.METHOD: Request.Method.GET,
                      RequestParser.FUNC: self._do_get_property,
                      RequestParser.PARAMS: 0})
        rules.append({RequestParser.PATTERN: r"^property/?$",
                      RequestParser.METHOD: Request.Method.PUT,
                      RequestParser.FUNC: self._do_put_property,
                      RequestParser.PARAMS: 1})
        rules.append({RequestParser.PATTERN: r"^settings/?$",
                      RequestParser.METHOD: Request.Method.GET,
                      RequestParser.FUNC: self._do_get_settings,
                      RequestParser.PARAMS: 0})
        rules.append({RequestParser.PATTERN: r"^settings/?$",
                      RequestParser.METHOD: Request.Method.PUT,
                      RequestParser.FUNC: self._do_put_settings,
                      RequestParser.PARAMS: 1})
        self.__parser.add_rule(rules)

    def dispatch_request(self, request):
        try:
            resp = self.__parser.action(request)
            if resp.status_code == Response.StatusCode.NOT_FOUND:
                resp = self._on_request(request)
        except:
            logging.error("Exception: Receive Request"
                          + " Method:" + request.method
                          + " Path:" + request.path)
            logging.error(traceback.format_exc())
            resp = Response(Response.StatusCode.INTERNAL_SERVER_ERROR, None)

        return resp

    def dispatch_event(self, event):
        self._do_post_event(event)

    def _do_get_property(self):
        logging.debug("Receive GET Property ObjectID:" + self.object_id)
        return Response(Response.StatusCode.OK, self._object_property)

    def _do_put_property(self, body):
        logging.debug("Receive PUT Property ObjectID:" + self.object_id)
        # check State Change to Finalizing
        for k, v in body.items():
            if k == ObjectProperty.OBJECT_STATE:
                oldValue = self._object_property.get_property(k)
                self.on_state_changed(oldValue, v)

        # update Property
        if not(self._object_property.equals(body)):
            self._object_property.put_property(body)

        return Response(Response.StatusCode.OK, self._object_property)

    def _do_get_settings(self):
        logging.debug("Receive GET Settings ObjectID:" + self.object_id)
        return Response(Response.StatusCode.OK, self._object_settings)

    def _do_put_settings(self, body):
        logging.debug("Receive PUT Settings ObjectID:" + self.object_id)
        if body != self._object_settings:
            old = copy.deepcopy(self._object_settings)
            # delete
            for k in old.keys():
                if k not in body:
                    del self._object_settings[k]

            # add or update
            for k, v in body.items():
                self._object_settings[k] = v

            self.on_settings_changed(ObjectSettingChanged.Action.UPDATE,
                                     old,
                                     self._object_settings)

        return Response(Response.StatusCode.OK, self._object_settings)

    def _do_post_event(self, event):
        self.on_event(event)

    def on_state_changed(self, old_state, new_state):
        if old_state != ObjectProperty.State.FINALIZING:
            if new_state == ObjectProperty.State.FINALIZING:
                self.on_finalize()

    def on_settings_changed(self, action, prev, curr):
        event_body = {ObjectSettingChanged.ACTION: action,
                      ObjectSettingChanged.PREV: prev,
                      ObjectSettingChanged.CURR: curr}
        self._publish_event_async(ObjectSettingChanged.TYPE,
                                  event_body)

    def _on_request(self, request):
        return Response(Response.StatusCode.BAD_REQUEST, None)

    def on_event(self, event):

        # for example: method_name = "_do_event_topologychanged"
        method_name = "_do_event_" + event.event_type.lower()
        if hasattr(self, method_name):
            method = getattr(self, method_name)
            if callable(method):
                method(event)
                return
        return

File: src/main/ruby/org/o3project/odenos/remoteobject/remote_object.rb

# Copyright 2015 NEC Corporation.                                          #
#                                                                          #
# Licensed under the Apache License, Version 2.0 (the "License");          #
# you may not use this file except in compliance with the License.         #
# You may obtain a copy of the License at                                  #
#                                                                          #
#   http://www.apache.org/licenses/LICENSE-2.0                             #
#                                                                          #
# Unless required by applicable law or agreed to in writing, software      #
# distributed under the License is distributed on an "AS IS" BASIS,        #
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. #
# See the License for the specific language governing permissions and      #
# limitations under the License.                                           #

require 'odenos/core/util/object_hash'
require 'odenos/remoteobject/event'
require 'odenos/remoteobject/request'
require 'odenos/remoteobject/response'
require 'odenos/core/util/logger'
require 'odenos/core/util/request_parser'

module Odenos
  module Core
    class ObjectProperty < Odenos::Util::ObjectHash
      ID = 'id'
      SUPER_TYPE = 'super_type'
      TYPE = 'type'
      STATE = 'state'
      DESCRIPTION = 'description'
      COMPONENT_TYPES = 'component_types'
      CONNECTION_TYPES = 'connection_types'

      hash_accessor :object, :state, :super_type
      hash_accessor :object, :description, :component_types, :connection_types

      def remote_object_type
        @object[:type]
      end

      def remote_object_type=(v)
        @object[:type] = v
      end

      def remote_object_id
        @object[:id]
      end

      def remote_object_id=(v)
        @object[:id] = v
      end

      alias_method :get_state, :state
      alias_method :set_state, :state=

      module State
        INITIALIZING = 'initializing'
        RUNNING = 'running'
        FINALIZING = 'finalizing'
        ERROR = 'error'
      end

      # @param [String,Symbol] key
      # @return old_value
      def set_property(key, value)
        key = key.to_sym if key.is_a? String

        if read_only?(key)
          return nil
        else
          old_value = @object[key]
          @object[key] = value
          return old_value
        end
      end

      # @param [String,Symbol] key
      def get_property(key)
        key = key.to_sym if key.is_a? String

        @object[key]
      end

      def delete_property(key)
        key = key.to_sym if key.is_a? String

        return if read_only?(key)
        @object.delete(key)
      end

      # @param [Hash] new_prop
      def put_property(new_prop)
        old_prop = clone
        # delete
        old_prop.to_hash.each_key do |key|
          unless new_prop.include?(key)
            delete_property(key)
          end
        end

        # add or update
        new_prop.each do |key, value|
          set_property(key, value)
        end
      end

      def initialize(initial_hash = {})
        super
      end

      # @param [Hash,ObjectProperty] other
      def ==(other)
        prop = {}
        if other.is_a? ObjectProperty
          prop = other.to_hash
        else
          other.each do |key, value|
            key = key.to_sym if key.is_a? String
            prop[key] = value
          end
        end
        @object == prop
      end

      def clone
        obj = ObjectProperty.new
        obj.remote_object_type = remote_object_type
        obj.remote_object_id = remote_object_id
        obj.super_type = super_type
        obj.description = description
        obj.connection_types = connection_types
        @object.each do |key, value|
          obj.set_property(key, value)
        end
        obj
      end

      def read_only?(key)
        if [:type, :id, :super_type, :description, :connection_types].include?(key)
          return true
        end
        false
      end
    end

    class RemoteObject
      include Odenos::Util
      include Odenos::Util::Logger

      # @return [EventSubscription]
      attr_accessor :event_subscription

      # @return [ObjectProperty]
      attr_accessor :property

      # @return [Hash]
      attr_accessor :settings

      # @return [MessageDispatcher]
      attr_reader :dispatcher

      # @!attribute [r] remote_object_id
      #   @return [String] ODENOS object id
      def remote_object_id
        property.remote_object_id
      end

      # @!attribute [r] state
      def state
        property.state
      end

      def on_initialize
        true
      end

      def on_finalize
        prev = property.clone
        if property.state != ObjectProperty::State::FINALIZING
          property.state = ObjectProperty::State::FINALIZING
          on_property_changed('delete', prev, nil)
        end
        dispatcher.remove_local_object(self)
      end

      def set_state(next_state)
        if property.state == next_state
          return
        end
        prev = property.clone
        property.state = next_state
        on_property_changed('update', prev, property)
      end

      # @param [String] remote_object_id
      # @param [Symbol] method One of :GET, :PUT, :POST, :DELETE
      # @param [String] path
      # @param [#to_msgpack] body
      # @return [Response]
      def request_sync(remote_object_id, method, path, body)
        @dispatcher.request_sync(Request.new(remote_object_id,
                                             method, path, body))
      end

      # @param [String] remote_object_id
      # @param [Symbol] method One of :GET, :PUT, :POST, :DELETE
      # @param [String] path
      # @param [#to_msgpack] body
      # @return [Response]
      def request(remote_object_id, method, path, body)
        debug { "request(#{remote_object_id},#{method},#{path},#{body.inspect})" }
        request_sync(remote_object_id, method, path, body)
      end

      # @param [String] event_type
      # @param [#to_msgpack] body
      def publish_event_async(event_type, body)
        @dispatcher.publish_event_async(Event.new(remote_object_id,
                                                  event_type, body))
      end

      #
      # Subscribe events of other RemoteObjects.
      # Caller should configure `event_subscription` member to
      # specify publisher IDs and events
      # before calling this method.
      # @return [Response]
      def apply_event_subscription
        @dispatcher.subscribe_event(event_subscription)
      end

      # @return [void]
      def subscribe_event
        apply_event_subscription
      end

      # @param [Request] request
      def dispatch_request(request)
        debug { "dispatch_request( #{request.inspect} )" }
        resp = nil
        begin
          resp = @parser.action(request)
          if resp.status_code == Response::NOT_FOUND
            resp = on_request(request)
          end
        rescue => ex
          resp = Response.new(Response::INTERNAL_SERVER_ERROR, nil)
          error "Exception: Receive Request : #{request.inspect}"
          error "#{ex.message} #{ex.backtrace}"
        end

        return resp
      ensure
        debug { "Exit dispatch_request( #{request.inspect} )" }
      end

      def dispatch_event(event)
        debug { "dispatch_event( #{event.inspect} )" }
        do_post_event(event)
      end

      def finalize?(old_state, new_state)
        if (old_state != ObjectProperty::State::FINALIZING)
          if (new_state == ObjectProperty::State::FINALIZING)
            return true
          end
        end
        false
      end

      def on_property_changed(action, prev, curr)
        publish_event_async(ObjectPropertyChanged::TYPE,
                            'action' => action,
                            'prev' => prev,
                            'curr' => curr)
      end

      def on_settings_changed(prev, curr)
        publish_event_async(ObjectSettingsChanged::TYPE,
                            'action' => 'update',
                            'prev' => prev,
                            'curr' => curr)
      end

      # @param [Request] request
      # @return [Response]
      def on_request(request)
        debug { "on_request( #{request.inspect} )" }
        Response.new(Response::BAD_REQUEST, '')
      end

      # @param [Event] _event
      # @return [void]
      def on_event(_event)
      end

      # @param [String] remote_object_id
      # @param [MessageDispatcher] dispatcher
      def initialize(remote_object_id, dispatcher)
        logger_ident_initialize(remote_object_id)
        debug "RemoteObject#initialize #{remote_object_id}"
        @property = ObjectProperty.new(type: self.class.to_s.split('::').last,
                                       id: remote_object_id)
        @property.set_state(ObjectProperty::State::INITIALIZING)
        @property.description = @description
        @property.connection_types = @connection_types
        @settings = {}
        @dispatcher = dispatcher
        @dispatcher.add_local_object self
        @event_subscription =
            EventSubscription.new(subscriber_id: remote_object_id)
        @parser = RequestParser.new
        add_rules
      end

      protected

      def add_rules
        rules = []
        rules.push(RequestParser::PATTERN => /^property$/,
                   RequestParser::METHOD => :GET,
                   RequestParser::FUNC => method(:do_get_property),
                   RequestParser::PARAMS => 0)
        rules.push(RequestParser::PATTERN => /^property$/,
                   RequestParser::METHOD => :PUT,
                   RequestParser::FUNC => method(:do_put_property),
                   RequestParser::PARAMS => 1)
        rules.push(RequestParser::PATTERN => /^settings$/,
                   RequestParser::METHOD => :GET,
                   RequestParser::FUNC => method(:do_get_settings),
                   RequestParser::PARAMS => 0)
        rules.push(RequestParser::PATTERN => /^settings$/,
                   RequestParser::METHOD => :PUT,
                   RequestParser::FUNC => method(:do_put_settings),
                   RequestParser::PARAMS => 1)
        @parser.add_rule(rules)
      end

      # @return [Response]
      def do_get_property
        Response.new(Response::OK, property)
      end

      # @param [Request] request
      # @return [Response]
      def do_put_property(request)
        # check state change
        request.body.each do |key, value|
          next if key != ObjectProperty::STATE

          old_state = property.get_state
          if self.finalize?(old_state, value)
            on_finalize
            return Response.new(Response::OK, property)
          end
        end

        # update property
        unless property == request.body
          prev = property.clone
          property.put_property(request.body)
          # publish ObjectPropertyChanged
          on_property_changed('update', prev, property)
        end

        Response.new(Response::OK, property)
      end

      # @return [Response]
      def do_get_settings
        Response.new(Response::OK, settings)
      end

      # @param [Request] request
      # @return [Response]
      def do_put_settings(request)
        if settings != request.body
          prev = settings.clone

          # delete
          prev.each_key do |key|
            unless request.body.include?(key)
              settings.delete(key)
            end
          end

          # add or update
          request.body.each do |key, value|
            settings[key] = value
          end

          # publish ObjectSettingChanged
          on_settings_changed(prev, settings)

        end
        Response.new(Response::OK, settings)
      end

      # @param [Event] event
      # @return [Response]
      def do_post_event(event)
        on_event(event)
        Response.new(Response::ACCEPTED, nil)
      end
    end
  end
end
